engine/plugins/client/client.js

```diff
@@ -6,37 +6,90 @@
 'use strict';
 
 const debug = require('debug')('bibbox:CLIENT:main');
-const clients = {};
+const redis = require('redis');
+const Q = require('q');
 
 /**
  * The Client object.
  *
+ * @param redisConfig
+ *   Redis configuration.
+ * @param persistent
+ *   If true redis will be use else in-memory store (mostly useful for testing).
+ *
  * @constructor
  */
-const Client = function Client() {};
+const Client = function Client(redisConfig, persistent = false) {
+    this.persistent = persistent;
+    this.clients = [];
+
+    if (this.persistent) {
+        // Extend configuration object with an connection plan.
+        redisConfig.retry_strategy = function(options) {
+            if (options.error && options.error.code === 'ECONNREFUSED') {
+                return new Error('The server refused the connection');
+            }
+            if (options.total_retry_time > 1000 * 60 * 60) {
+                return new Error('Retry time exhausted');
+            }
+            if (options.attempt > 10) {
+                return undefined;
+            }
+            // Reconnect after.
+            return Math.min(options.attempt * 100, 3000);
+        };
+
+        this.storage = redis.createClient(redisConfig);
+    }
+};
 
 /**
  * Load client.
  *
  * @param token
  *   The token for the client.
+ * @param config
+ *   Configuration for FBS.
+ * @param state
+ *   The current state.
+ *
  * @return {*}
  *   The client for the token.
  */
-Client.prototype.load = (token, config = {}, state = {}) => {
-    // If client does not exists
-    if (!Object.prototype.hasOwnProperty.call(clients, token)) {
-        clients[token] = {
-            token: token,
-            config: config,
-            state: state
-        };
-        debug('Created new client: ' + token);
+Client.prototype.load = function load(token, config = {}, state = {}) {
+    const deferred = Q.defer();
+    let client = false;
+
+    if (this.persistent) {
+        this.storage.get(token, function get(err, data) {
+            if (err) {
+                deferred.resolve({
```

Comment:
Is this a new client (then add a comment), or should it be reject?

Reply:
Debug statements corrected and moved above.